### ISSUE TITLE
Make match_same_arms lint fail late

### ISSUE THREAD

**Issue by flip1995** [rust-lang/rust-clippy#4096]
Currently the `match_same_arms` lint fails early. If we have code like this:
```rust
match a {
    Foo::A => 1,
    Foo::B => 1,
    Foo::C => 2,
    Foo::D => 2,
};
```
the lint first triggers on `Foo::A` and `Foo::B` and when this case is fixed it will trigger on `Foo::C` and `Foo::D`. This is especially annoying on big `match`es where there are multiple identical arms.

It would be nice if the lint would first collect all the identical tuples of arms and then warn on all of them at once, instead of failing early.

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81c99a8499564bb2f8bdea11c42bfb70)
